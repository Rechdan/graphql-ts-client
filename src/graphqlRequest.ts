import _axios, { AxiosStatic } from 'axios'
import { ClientConfig, GraphQLClientError } from './types'

export async function graphqlRequest({
  axios = _axios,
  queryName,
  client,
  query,
  variables,
}: {
  axios?: AxiosStatic
  client: ClientConfig
  queryName: string
  query: string
  variables: { [_key: string]: any }
}) {
  let lastResponse: {
    data: any
    errors: { message: string }[]
    warnings: any[]
    headers: any
    status?: number
  }

  for (let trial = 0; trial <= client.retryConfig.max; trial++) {
    const infoParams = {
      _q: queryName,
      ...(trial > 0 ? { _retrial: trial + 1 } : {}),
    }

    const {
      data: responseData = {} as any,
      headers,
      status,
    } = await axios.post(
      client.url,
      { query, variables },
      {
        params: infoParams,
        headers: {
          'Content-Type': 'application/json',
          ...client.headers,
        },
        validateStatus: () => true,
      }
    )

    let { data, errors, warnings } = responseData

    if (status >= 400 && !errors?.length) {
      errors = [{ message: `Request "${queryName}" failed with status ${status}` }]
    }

    lastResponse = {
      errors,
      data,
      warnings,
      headers,
      status,
    }

    if (!errors?.length) {
      break
    } else if (trial < client.retryConfig.max && typeof client.retryConfig?.before === 'function') {
      await client.retryConfig?.before({
        queryName,
        query,
        variables,
        response: lastResponse,
      })
    }
  }

  if (lastResponse!.errors && lastResponse!.errors?.length) {
    throw new GraphQLClientError(lastResponse!.errors)
  }

  return lastResponse!
}
