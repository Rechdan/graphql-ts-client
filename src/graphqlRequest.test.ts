import { graphqlRequest } from './graphqlRequest'

describe('GraphQLRequest', () => {
  it('Should retry as many times as configured properly running a "before" hook', async () => {
    let retryCount = 0
    const maxRetrials = 2
    const mockedAxios = {
      post: async () => ({
        status: retryCount < maxRetrials ? 500 : 200,
      }),
    } as any

    const result = await graphqlRequest({
      failureMode: 'loud',
      axios: mockedAxios,
      queryName: 'whatever',
      query: 'whatever',
      variables: {},
      client: {
        url: 'https://whatever.com',
        headers: {},
        retryConfig: {
          max: maxRetrials,
          before: () => void [retryCount++],
        },
      },
    })

    expect(retryCount).toBe(2)
    expect(result.status).toBe(200)
  })

  it('Should ignore default retrying config explicitly asking to skip retrials ', async () => {
    let retryCount = 0
    const maxRetrials = 2
    const mockedAxios = {
      post: async () => ({
        status: retryCount < maxRetrials ? 500 : 200,
      }),
    } as any

    const result = await graphqlRequest({
      shouldRetry: false,
      failureMode: 'loud',
      axios: mockedAxios,
      queryName: 'whatever',
      query: 'whatever',
      variables: {},
      client: {
        url: 'https://whatever.com',
        headers: {},
        retryConfig: {
          max: maxRetrials,
          before: () => void [retryCount++],
        },
      },
    }).catch(err => err)

    expect(result).toBeInstanceOf(Error)
  })
})
