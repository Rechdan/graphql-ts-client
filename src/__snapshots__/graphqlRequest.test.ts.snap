// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GraphQLRequest Should request have proper structure 1`] = `
Arguments [
  "https://whatever.com",
  {
    "operationName": "sampleQueryName",
    "query": "sampleQuery",
    "variables": {
      "bar": "foo",
      "foo": "bar",
    },
  },
  {
    "headers": {
      "Content-Type": "application/json",
    },
    "params": {
      "_q": "sampleQueryName",
    },
    "validateStatus": [Function],
  },
]
`;
